### ISSUE TITLE
[BUG] Log axis bounds calculated incorrectly with certain data sets since v2.2.0

### ISSUE THREAD

**Issue by schleising** [bokeh/bokeh#10511]
#### ALL software version info (bokeh, python, notebook, OS, browser, any other relevant packages)
python 3.8.5 64 bit
bokeh 2.2.0, bokeh 2.2.1 and bokeh 2.3.0.dev1
pandas 1.1.2
Mac OS 10.15.6
Chrome Version 85.0.4183.102 (Official Build) (64-bit)
Safari Version 14.0 (15610.1.28.1.9, 15610)
Docker container: python:3.8-slim-buster

#### Description of expected behavior and the observed behavior

In bokeh up to and including v2.1.1 the log axes are calculated correctly.
In bokeh v2.2.0 and onwards the log axes are calculated incorrectly and a warning is issued in the Chrome Javascript console

> bokeh-2.2.1.min.js:387 [bokeh] could not determine minimum data value for log axis, DataRange1d using value 0.04

#### Complete, minimal, self-contained example code that reproduces the issue

This is the simplest data set I could produce to trigger the error

1. If you replace the 0.0 in y_vals with any value larger than 0 then the error does not appear to occur.
2. If you remove any x, y data pair then the error does not appear to occur

```
import pandas as pd

from bokeh.plotting import figure, ColumnDataSource, show

# Construct a data frame from the data which causes the issue
cols = {'x_vals': [11,11,11,11,12,12,12,12,12,12,12,12,12,12,14,18,38],
        'y_vals': [1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,1.0,0.0,1.0,4.0]}

new_df = pd.DataFrame(cols)

# Create the figure
plot = figure(title         = 'x-vals against y-vals',
              x_axis_label  = 'x-vals',
              y_axis_label  = 'y-vals',
              x_axis_type   = 'log',
              y_axis_type   = 'log',
              tools         = 'pan, wheel_zoom, reset',
              active_drag   = 'pan',
              active_scroll = 'wheel_zoom',
              sizing_mode   = 'stretch_both')

# Create a ColumnDataSource from the data frame
cds = ColumnDataSource(new_df)

# Add the line plot
p = plot.line(x          = 'x_vals',
              y          = 'y_vals',
              source     = cds,
              name       = 'Test',
              line_color = 'red',
              line_width = 1)

# Show the chart
show(plot)
```

#### Stack traceback and/or browser JavaScript console output
[bokeh] could not determine minimum data value for log axis, DataRange1d using value 0.04
_compute_range @ bokeh-2.2.1.min.js:387
async function (async)
lazy_initialize @ bokeh-2.2.1.min.js:495
i @ bokeh-2.2.1.min.js:284
t.build_view @ bokeh-2.2.1.min.js:284
r @ bokeh-2.2.1.min.js:572
t.add_document_standalone @ bokeh-2.2.1.min.js:572
O @ bokeh-2.2.1.min.js:164
(anonymous) @ bokeh-2.2.1.min.js:164
(anonymous) @ bokeh-2.2.1.min.js:185
requestAnimationFrame (async)
(anonymous) @ bokeh-2.2.1.min.js:185
t.defer @ bokeh-2.2.1.min.js:185
t.embed_items @ bokeh-2.2.1.min.js:164
embed_document @ test_log.html:55
(anonymous) @ test_log.html:59
(anonymous) @ test_log.html:75
o.safely @ bokeh-2.2.1.min.js:583
fn @ test_log.html:49

#### Screenshots or screencasts of the bug in action
When the axes are correctly calculated, the chart looks like this (using v2.1.1 and earlier),
<img width="1293" alt="Screenshot 2020-09-19 at 14 35 19" src="https://user-images.githubusercontent.com/22272455/93668630-82300180-fa85-11ea-928e-1618b02483f8.png">

When the error occurs the chart looks like this (using 2.2.0 and later) ,
<img width="1294" alt="Screenshot 2020-09-19 at 14 35 31" src="https://user-images.githubusercontent.com/22272455/93668645-a2f85700-fa85-11ea-93bb-c5732717a6b2.png">

**Comment by bryevdv** [comment]
@mattpap bisection results say this was broken in 7a03120c49463e40a44f41521a80d4389a270281 

> 7a03120c49463e40a44f41521a80d4389a270281 is the first bad commit
> commit 7a03120c49463e40a44f41521a80d4389a270281
> Author: Mateusz Paprocki <mattpap@gmail.com>
> Date:   Fri Jun 19 00:37:16 2020 +0200
> 
>     Reduce memory footprint of data indexing (#10162)
> 
>     * Reduce memory footprint of data indexing
> 
>     * Reimplement Patches/MultiPolygons indexing
> 
>  bokehjs/src/lib/core/util/spatial.ts            |  60 ++++++++++---
>  bokehjs/src/lib/models/glyphs/bezier.ts         |  20 ++---
>  bokehjs/src/lib/models/glyphs/box.ts            |  20 ++---
>  bokehjs/src/lib/models/glyphs/glyph.ts          |  63 +++++++++-----
>  bokehjs/src/lib/models/glyphs/harea.ts          |  17 ++--
>  bokehjs/src/lib/models/glyphs/hbar.ts           |   5 --
>  bokehjs/src/lib/models/glyphs/hex_tile.ts       |  16 ++--
>  bokehjs/src/lib/models/glyphs/image_base.ts     |  16 ++--
>  bokehjs/src/lib/models/glyphs/image_url.ts      |   9 +-
>  bokehjs/src/lib/models/glyphs/multi_line.ts     |  19 +++--
>  bokehjs/src/lib/models/glyphs/multi_polygons.ts | 109 +++++++++++++++++-------
>  bokehjs/src/lib/models/glyphs/patches.ts        |  27 +++---
>  bokehjs/src/lib/models/glyphs/quad.ts           |   5 --
>  bokehjs/src/lib/models/glyphs/quadratic.ts      |  19 ++---
>  bokehjs/src/lib/models/glyphs/segment.ts        |  22 ++---
>  bokehjs/src/lib/models/glyphs/varea.ts          |  17 ++--
>  bokehjs/src/lib/models/glyphs/vbar.ts           |   5 --
>  bokehjs/src/lib/models/glyphs/xy_glyph.ts       |  16 ++--
>  bokehjs/test/unit/models/glyphs/glyph.ts        |   4 +-
>  19 files changed, 268 insertions(+), 201 deletions(-)

**Comment by bryevdv** [comment]
That said the proximate cause seems to be that `this._initial_start` is zero

**Comment by bryevdv** [comment]
@schleising I actually looked over this issue too quickly and missed the zero value. That is not a mathematically valid value for a log axis, (the "correct" bounds wold be negative infinity) I am not sure what behavior you are expecting but I think the correct course of action regardless is for users to ensure that there are never zero values on a log scale.

**Comment by bryevdv** [comment]
@mattpap that said, the changes in https://github.com/bokeh/bokeh/commit/7a03120c49463e40a44f41521a80d4389a270281 did cause lines with `np.nan` on a log scale to break bounds computation, and I *do* think that should work If the zero in the example is repaced with nan

* nan, linear scale (good):

<img width="424" alt="Screen Shot 2020-09-19 at 1 08 55 PM" src="https://user-images.githubusercontent.com/1078448/93688254-5149cf80-fa79-11ea-8af1-fbe278b40977.png">

* nan, log scale (bad):

<img width="421" alt="Screen Shot 2020-09-19 at 1 09 05 PM" src="https://user-images.githubusercontent.com/1078448/93688252-4f800c00-fa79-11ea-906f-55f0907611fa.png">

**Comment by schleising** [comment]
> @schleising I actually looked over this issue too quickly and missed the zero value. That is not a mathematically valid value for a log axis, (the "correct" bounds wold be negative infinity) I am not sure what behavior you are expecting but I think the correct course of action regardless is for users to ensure that there are never zero values on a log scale.

Hi @bryevdv 

Thanks for looking into the issue so quickly, I'm aware that log(0) is not a valid calculation and have no problem dropping zeros from the data frame, I've already given this a try on the full data set and it works fine.

The only reason I raised it as a bug is that previous versions handled the situation gracefully and the newer versions don't.  I'm not sure what the philosophy you have here is, so I'm happy to go along with whatever you think is the right thing to do.

If you decide that the user should ensure there are no 0 values in a log plot, is it worth considering a more informative error/warning message?

Thanks for the work you're doing on Bokeh, I've used it as my introduction to Python and it's been a great tool to use to learn how to use the language.

**Comment by ahesford** [comment]
Commit 7a03120c49463e40a44f41521a80d4389a270281 has bitten me in a different way when using a log scale on a `vbar` plot, even when there are no zeros in the data. I'm running a Bokeh server application that initially creates an empty `vbar` plot with a `ColumnDataSource` to populate the `x` and `top` properties of the plot. Both columns are empty upon initialization of the plot, and I initialize the `bottom` and `width` properties each to 1. When I update the data source with non-empty columns, I adjust the width to some arbitrary but positive value and attempt to set the `bottom` to a value lower than the smallest value in the `top` column but not less than 1 decade below this minimum value.

In earlier commits, this has the effect of showing a bar for every value with a nonzero height. In 7a03120c49463e40a44f41521a80d4389a270281 and later, the bars behave as expected *except* that the vertical axis of the plot is inverted and the automatic scaling is not correctly calculated.

Images are attached. I will try to get a minimal demonstration when possible.

Default view with autoscale, commit 7a03120
![7a0312-autoscale](https://user-images.githubusercontent.com/48421688/95029623-75abcb80-0677-11eb-85c8-d0aff527262f.png)

Manually rescaled, commit 7a03120
![7a0312-manual](https://user-images.githubusercontent.com/48421688/95029624-76446200-0677-11eb-8b89-9ac81aa3ec68.png)

Default view with autoscale, commit f190c4a9
![f190c4-autoscale](https://user-images.githubusercontent.com/48421688/95029697-126e6900-0678-11eb-8412-164a23c72496.png)

**Comment by ahesford** [comment]
OK, this example may not be strictly minimal, but it is basic enough and reflects my use of `vbar` in a way that demonstrates the bugs introduced with 7a03120.

The `MagViewer` creates a plot with `y_axis_type='log'` and an empty `renderers` map from a string label to a pair of `ColumnDataSource` and the `GlyphRenderer` returned by `vbar` for that data source. Each `vbar` plot uses its data source to define the `x` and `top` properties of the glyphs, while the `bottom` and `width` properties are held constant over all bars.

`MagViewer._adjustBarData` accepts a label, a list of `x` column values, a list of `top` column values, a scalar `bottom` value and a scalar `width`. The label is used to retrieve an existing (`ColumnDataSource`, `GlyphRenderer`) pair if one already exists, setting the data for the `ColumnDataSource` to the `x` and `top` values provided and the `bottom` and `width` properties of the `GlyphRenderer` to those provided values. If a pair does not exist, `vbar` is used to create a new pair with equivalent properties, and these are retained for subsequent calls.

On initialization, the `MagViewer` uses `_adjustBarData` to prepopulate one renderer with an empty data source, then presents an `update` method (called when a button is pressed) that generates random arrays of `x` and `top` data, then iteratively invokes `_adjustBarData` a total of four times, where the bars drawn in iteration `i` represent every `(i + 1)`-th value of the random data.

Run the below example, saved as `mag.py`, with `bokeh serve --show mag.py` and click the "Press Me" button. Note that the plot should automatically scale, but the scaling is not correct. Also note that the y axis increases downward, which is unexpected and different from prior behavior. This has been tested in Firefox and qutebrowser (which uses qt5-webengine as a backend).

[View or save script mag.py.txt](https://github.com/bokeh/bokeh/files/5324802/mag.py.txt)

**Comment by ahesford** [comment]
Ping.

The issue still affects the newly-released 2.3. I wonder if anybody has been able to reproduce these issues and if there is anything else I can do to provide some insight.

**Comment by bryevdv** [comment]
> The issue still affects the newly-released 2.3. I

It's slated for the 2.4 milestone

**Comment by mattpap** [comment]
This was fixed in PR #11510.

**Comment by bryevdv** [comment]
duplicate of #11508
